python/tvm/contrib/torch/as_torch.py

```diff
@@ -0,0 +1,88 @@
+# pylint: disable=inconsistent-return-statements
+#!/usr/bin/env python
+
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+# pylint: disable=missing-module-docstring
+# pylint: disable=missing-class-docstring
+# pylint: disable=missing-function-docstring
+"""
+as_torch: a decorator, which is used to wrap the TVMscript code to `torch.nn.module`.
+"""
+from typing import Callable, List, Union
+
+import torch
+import torch.utils.dlpack
+
+import tvm
+
+
+# python wrapper for OperatorModule
+class OperatorModuleWrapper(torch.nn.Module):
+    def __init__(
+        self,
+        module: Union[
+            tvm.ir.module.IRModule,
+            tvm.tir.function.PrimFunc,
+        ],
+    ):
+        super().__init__()
+        self.rt_module = None  # runtime module
+        self.ir_module = module  # IR modules
+
+    def build(self, target=None):
+        runtime_module = tvm.build(self.ir_module, target=target)
+        func = tvm.get_global_func("tvmtorch.save_runtime_mod")
+        func(runtime_module)
+
+        self.rt_module = torch.classes.tvm_torch.OperatorModuleWrapper()
+
+    def forward(self, *torch_inputs: List[torch.Tensor]) -> List[torch.Tensor]:
+        if self.rt_module is None:
+            if torch_inputs[0].is_cuda:
+                self.build(target="cuda")
+            elif torch_inputs[0].device.type == "cpu":
+                self.build()
+            else:
+                raise Exception(f"the target {torch_inputs[0].device.type} is not supported yet")
+
+        return self.rt_module.forward(torch_inputs)
+
+
+def as_torch(func: Union[tvm.ir.module.IRModule, tvm.tir.function.PrimFunc, Callable]):
```

Comment:
So `as_torch` doesn't provide tuning facilities? I noticed that all tuning tests in this PR is done via `optimize_torch` which involves Relay. If a user wants to tune a TVMScript-written op and use `@as_torch` decorator, how tuning can be done?

Reply:
It's still possible for PT users to write TVMScript and `tune_tir` to tune, and use `as_torch` to convert the tuned prim func to PT. We are offering `optimize_pytorch` to wrap `tune_relay`, so it would be nice if `as_torch` also wraps `tune_tir` and automatically does tuning.

Current examples only show the usage of `as_torch` as an decorator on top of a manually written TVMScript without tuning.

Reply:
How about this: We add `tune(config) method on `OperatorModuleWrapper`, which does tuning and rebuild the mod.  And remove `tune_tir` from `build(...)`. So by default tuning won't happen, but the user can explicitly ask to tune.